unic/char/range/src/iter.rs

```diff
@@ -0,0 +1,112 @@
+use std::{char, iter};
+use CharRange;
+
+/// Range of Surrogate Code Points.
+///
+/// Reference: <http://unicode.org/glossary/#surrogate_code_point>
+const SURROGATE_RANGE: ::std::ops::Range<u32> = 0xD800..0xE000;
+
+/// An iterator over `char`.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct CharIter {
+    /// The codepoint of the smallest uniterated codepoint.
+    ///
+    /// If greater than or equal to `high`, iteration is finished.
+    ///
+    /// # Safety
+    ///
+    /// Must be a valid, non-surrogate codepoint while iteration is unfinished.
+    low: u32,
+
+    /// The codepoint one greater than the largest uniterated codepoint.
+    ///
+    /// If less than or equal to `low`, iteration is finished.
+    ///
+    /// # Safety
+    ///
+    /// Must be one greater than a valid, non surrogate codepoint while iteration is unfinished.
+    high: u32,
+}
+
+impl<'a> From<&'a CharRange> for CharIter {
+    fn from(range: &CharRange) -> CharIter {
+        CharIter {
+            low: range.first() as u32,
+            high: range.last() as u32 + 1,
+        }
+    }
+}
+
+impl CharIter {
+    #[inline]
+    fn is_finished(&self) -> bool {
+        self.low >= self.high
+    }
+}
+
+impl Iterator for CharIter {
+    type Item = char;
+
+    #[inline]
+    #[allow(unsafe_code)]
+    fn next(&mut self) -> Option<char> {
+        if self.is_finished() {
+            return None;
+        }
+
+        let char = unsafe { char::from_u32_unchecked(self.low) };
```

Comment:
👍 it's a bad habit. It's not something you'd really expect to work, but it does (I think because Rust does type and variable name lookup differently). Surprisingly though, my IDE highlights it right, which is part of why I do it even though I probably shouldn't.